File: app/src/test/java/com/witype/Dragger/mvp/present/HomePresenterTest.java
package com.witype.Dragger.mvp.present;

import com.witype.Dragger.RxJavaRule;
import com.witype.Dragger.entity.MobileDateUsageEntity;
import com.witype.Dragger.entity.RecordsBean;
import com.witype.Dragger.entity.ResultBean;
import com.witype.Dragger.integration.HttpModel;
import com.witype.Dragger.integration.HttpObserver;
import com.witype.Dragger.integration.UIObservableTransformer;
import com.witype.Dragger.mvp.contract.HomeView;
import com.witype.Dragger.mvp.model.ApiModel;
import com.witype.mvp.integration.RetrofitRequestManager;

import org.hamcrest.Matchers;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InOrder;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.junit.MockitoJUnitRunner;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import io.reactivex.Observable;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;
import io.reactivex.observers.TestObserver;
import io.rx_cache2.internal.RxCache;
import io.victoralbertos.jolyglot.GsonSpeaker;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

import static org.junit.Assert.assertThat;

@RunWith(MockitoJUnitRunner.class)
public class HomePresenterTest {

    public static final int LIMIT = 100;

    @Rule
    public RxJavaRule rule = new RxJavaRule();

    @Mock
    HomePresenter homePresenter;

    @Mock
    List<RecordsBean> mData;

    @Mock
    MobileDateUsageEntity mobileDateUsageEntity;

    Random random;

    @Before
    public void setUp() throws Exception {
        HomeView homeView = Mockito.mock(HomeView.class);
        HttpModel httpModel = Mockito.mock(HttpModel.class);

        Mockito.doReturn(homeView).when(homePresenter).getView();
        Mockito.doReturn(httpModel).when(homePresenter).getModel();

        random = new Random();
        mData = new ArrayList<>();
        int year = random.nextInt(2000);
        for (int i = 1; i <= LIMIT; i++) {
            RecordsBean spy = new RecordsBean();
            int quarter = Math.max(random.nextInt(4), 1);
            spy.setQuarter(String.format("%s Q%s", year, quarter));
            spy.setVolume_of_mobile_data(random.nextFloat());
            mData.add(spy);
            year = i % 4 == 0 ? random.nextInt(2000) : year;
        }

        // mock
        Observable<MobileDateUsageEntity> just = Observable.just(mobileDateUsageEntity);
        ResultBean resultBean = Mockito.mock(ResultBean.class);
        Mockito.when(resultBean.getRecords()).thenReturn(mData);
        Mockito.when(mobileDateUsageEntity.getResult()).thenReturn(resultBean);
        Mockito.when(homePresenter.getModel().getMobileDataUsage(HomePresenter.RESOURCE_ID,LIMIT)).thenReturn(just);
    }

    /**
     * 测试调用 {@link HomeView#onGetMobileDataUsage()}
     * use RxJava Observer {@link HttpObserver}
     */
    @Test
    public void testHomeViewCallSuccessWithHttpObservable() {
        log("┌─── test testHomeViewCallSuccess ───────────────────────────────────────");
        HttpObserver<MobileDateUsageEntity> httpObserver = new HttpObserver<MobileDateUsageEntity>() {
            @Override
            public void onSuccess(MobileDateUsageEntity mobileDateUsageEntity) {
                log("|\t call onSuccess => onGetMobileDataUsage");
                homePresenter.getView().onGetMobileDataUsage();
            }

            @Override
            public void onError(Throwable e) {
                super.onError(e);
                log("|\t call onError => onGetDataError");
                homePresenter.getView().onGetDataError(e.getMessage());
            }
        };
        TestObserver<MobileDateUsageEntity> testObserver = new TestObserver<MobileDateUsageEntity>(httpObserver);

        homePresenter.getModel()
                .getMobileDataUsage(HomePresenter.RESOURCE_ID, LIMIT)
                .subscribe(testObserver);
        testObserver.assertNoErrors();

        log("|\t▊ assert");
        log("|\t assert HomeView#onGetMobileDataUsage atMostOnce call");
        Mockito.verify(homePresenter.getView(),Mockito.atMostOnce()).onGetMobileDataUsage();
        log("|\t assert HomeView#onGetDataError never call");
        Mockito.verify(homePresenter.getView(),Mockito.never()).onGetDataError(Mockito.any());
        log("└──────────────────────────────────────────────────────────────");
    }

    /**
     * 测试调用 {@link HomeView#onGetDataError(String)} ()}
     * use RxJava Observer {@link HttpObserver}
     */
    @Test
    public void testHomeViewCallFailureModelErrorWithHttpObservable() {
        log("┌─── test testHomeViewCallFailureModelErrorWithHttpObservable ────────────────────");
        //mock data
        Mockito.when(mobileDateUsageEntity.getResult().getRecords()).thenReturn(null);

        HttpObserver<MobileDateUsageEntity> httpObserver = new HttpObserver<MobileDateUsageEntity>() {
            @Override
            public void onSuccess(MobileDateUsageEntity mobileDateUsageEntity) {
                log("|\t mock nullpointException");
                int size = mobileDateUsageEntity.getResult().getRecords().size();
                homePresenter.getView().onGetMobileDataUsage();
            }

            @Override
            public void onError(Throwable e) {
                super.onError(e);
                log("|\t onError : " + e.getMessage());
                homePresenter.getView().onGetDataError(e.getMessage());
            }
        };
        TestObserver<MobileDateUsageEntity> testObserver = new TestObserver<MobileDateUsageEntity>(httpObserver);

        homePresenter.getModel()
                .getMobileDataUsage(HomePresenter.RESOURCE_ID, LIMIT)
                .subscribe(testObserver);
        testObserver.assertNoErrors();

        log("|\t▊ assert");
        log("|\t assert HomeView#onGetMobileDataUsage never call ");
        Mockito.verify(homePresenter.getView(),Mockito.never()).onGetMobileDataUsage();
        log("|\t assert HomeView#onGetDataError atMostOnce call");
        Mockito.verify(homePresenter.getView(),Mockito.atMostOnce()).onGetDataError(Mockito.any());

        log("└────────────────────────────────────────────────────");
    }

    /**
     * 测试
     * 1、{@link UIObservableTransformer#hasSubscribe()}，当订阅成功后调用，
     * 2、{@link UIObservableTransformer#willComplete()} , 当订阅完成后调用
     * 在UI使用中，可以通过以上两个方法同步界面，如：网络加载时显示等待框，请求完成时隐藏等贷框
     */
    @Test
    public void testUIObservableTransformer() {
        log("┌─── test testUIObservableTransformer ─────────────────────────────────");

        HttpObserver<MobileDateUsageEntity> httpObserver = new HttpObserver<MobileDateUsageEntity>() {
            @Override
            public void onSuccess(MobileDateUsageEntity mobileDateUsageEntity) {
                log("|\t call onSuccess => onGetMobileDataUsage");
                homePresenter.getView().onGetMobileDataUsage();
            }

            @Override
            public void onError(Throwable e) {
                super.onError(e);
                log("|\t call onError => onGetDataError");
                homePresenter.getView().onGetDataError(e.getMessage());
            }
        };
        TestObserver<MobileDateUsageEntity> testObserver = new TestObserver<MobileDateUsageEntity>(httpObserver);

        homePresenter.getModel()
                .getMobileDataUsage(HomePresenter.RESOURCE_ID, LIMIT)
                .compose(new UIObservableTransformer<MobileDateUsageEntity>() {
                    @Override
                    public void hasSubscribe() {
                        log("|\t call hasSubscribe => showLoading");
                        homePresenter.getView().showLoading();
                    }

                    @Override
                    public void willComplete() {
                        log("|\t call willComplete => dismissLoading");
                        homePresenter.getView().dismissLoading();
                    }
                })
                .subscribe(testObserver);
        testObserver.assertNoErrors();

        InOrder inOrder = Mockito.inOrder(homePresenter.getView());
        log("|\t┌─── assert ────────────────────────────────────────────────");
        log("|\t▊ assert order showLoading => onGetMobileDataUsage => dismissLoading");
        inOrder.verify(homePresenter.getView()).showLoading();
        inOrder.verify(homePresenter.getView()).onGetMobileDataUsage();
        inOrder.verify(homePresenter.getView()).dismissLoading();

        log("|\t▊ assert");
        log("|\t| assert HomeView#onGetMobileDataUsage atMostOnce call");
        Mockito.verify(homePresenter.getView(),Mockito.atMostOnce()).onGetMobileDataUsage();
        log("|\t| assert HomeView#showLoading atMostOnce call");
        Mockito.verify(homePresenter.getView(),Mockito.atMostOnce()).showLoading();
        log("|\t| assert HomeView#dismissLoading atMostOnce call");
        Mockito.verify(homePresenter.getView(),Mockito.atMostOnce()).dismissLoading();
        log("\t└────────────────────────────────────────────────────────────");

        log("└────────────────────────────────────────────────────");
    }

    /**
     * 测试{@link ApiModel#getMobileDataUsage(String, int)}https://data.gov.sg获取数据异常的情况，模拟{@link HomeView#onGetDataError(String)}
     * see <a>https://data.gov.sg/api/action/datastore_search?resource_id=a807b7ab-6cad-4aa6-87d0-e283a7353a0f&limit=5</a>
     */
    @Test
    public void testGetRemoteDataFailureWithHttpObserver() {
        log("┌─── test testGetRemoteDataFailureWithHttpObserver ────────────────────");
        HttpObserver<MobileDateUsageEntity> httpObserver = new HttpObserver<MobileDateUsageEntity>() {
            @Override
            public void onSuccess(MobileDateUsageEntity mobileDateUsageEntity) throws Exception {
                log("|\t mock exception");
                //mock exception
                throw new NullPointerException("mock exception");
            }

            @Override
            public void onResponseError(String message, int code) {
                super.onResponseError(message, code);
                homePresenter.getView().onGetDataError(message);
            }

            @Override
            public void onNetworkError(Throwable t) {
                super.onNetworkError(t);
                homePresenter.getView().onGetDataError(t.getMessage());
            }
        };
        TestObserver<MobileDateUsageEntity> testObserver = new TestObserver<MobileDateUsageEntity>(httpObserver);

        homePresenter.getModel()
                .getMobileDataUsage(HomePresenter.RESOURCE_ID, LIMIT)
                .compose(new UIObservableTransformer<MobileDateUsageEntity>() {
                    @Override
                    public void hasSubscribe() {
                        log("|\t call hasSubscribe => showLoading");
                        homePresenter.getView().showLoading();
                    }

                    @Override
                    public void willComplete() {
                        log("|\t call willComplete => dismissLoading");
                        homePresenter.getView().dismissLoading();
                    }
                })
                .subscribe(testObserver);
        testObserver.assertNoErrors().assertComplete();

        InOrder inOrder = Mockito.inOrder(homePresenter.getView());
        log("|\t┌─── assert ────────────────────────────────────────────────");
        log("|\t▊ assert order showLoading => onGetMobileDataUsage => dismissLoading");
        inOrder.verify(homePresenter.getView()).showLoading();
        inOrder.verify(homePresenter.getView()).onGetDataError("mock exception");
        inOrder.verify(homePresenter.getView()).dismissLoading();

        log("|\t▊ assert call");
        log("|\t| assert HomeView#showLoading atMostOnce call");
        Mockito.verify(homePresenter.getView(),Mockito.atMostOnce()).showLoading();
        log("|\t| assert HomeView#onGetDataError atMostOnce call");
        Mockito.verify(homePresenter.getView(),Mockito.atMostOnce()).onGetDataError("mock exception");
        log("|\t| assert HomeView#dismissLoading atMostOnce call");
        Mockito.verify(homePresenter.getView(),Mockito.atMostOnce()).dismissLoading();
        log("\t└────────────────────────────────────────────────────────────");

        log("└────────────────────────────────────────────────────");
    }

    private void log(String message) {
        System.out.println(message);
    }
}

File: kotlindemo/src/test/java/com/witype/kotlindemo/mvp/present/HomePresenterTest.kt
package com.witype.kotlindemo.mvp.present

import com.google.gson.Gson
import com.witype.kotlindemo.RxJavaRule
import com.witype.kotlindemo.entity.MobileDateUsageEntity
import com.witype.kotlindemo.entity.RecordsBean
import com.witype.kotlindemo.entity.ResultBean
import com.witype.kotlindemo.integration.HttpModel
import com.witype.kotlindemo.integration.HttpObserver
import com.witype.kotlindemo.mvp.contract.HomeView
import com.witype.kotlindemo.mvp.model.ApiModel
import com.witype.kotlindemo.mvp.presenter.HomePresenter
import com.witype.mvp.integration.UIObservableTransformer
import io.mockk.*
import io.mockk.impl.annotations.MockK
import io.reactivex.Observable
import io.reactivex.android.plugins.RxAndroidPlugins
import io.reactivex.observers.TestObserver
import io.reactivex.plugins.RxJavaPlugins
import okhttp3.OkHttpClient
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import kotlin.random.Random


class HomePresenterTest {

    val LIMIT = 2018

    /**
     * 设置rxJava所有的执行线程为[io.reactivex.schedulers.Schedulers.trampoline]
     * set all rxjava scheduler run with [io.reactivex.schedulers.Schedulers.trampoline]
     */
    @get:Rule
    val rxJava: RxJavaRule = RxJavaRule()

    lateinit var mData: MutableList<RecordsBean>

    @MockK
    lateinit var model: HttpModel;

    lateinit var onGetMobileDataUsage: HomeView

    lateinit var onGetDataError: HomeView

    lateinit var showLoading: HomeView

    lateinit var dismissLoading: HomeView

    @MockK
    lateinit var mobileDateUsageEntity: MobileDateUsageEntity

    fun getYearRange(): Int {
        return (1000..2000).random()
    }

    @Before
    fun setup() {
        MockKAnnotations.init(this)
        mData = ArrayList()
        val count = (50..100).random()
        var year = getYearRange();
        for (i in 1..count) {
            val spy = RecordsBean()
            val quarter = (1..4).random().coerceAtLeast(1)
            spy.quarter = String.format("%s Q%s", year, quarter)
            spy.volume_of_mobile_data = Random.nextDouble()
            mData.add(spy)
            year = if (i % 4 == 0) getYearRange() else year
        }

        //mock
        val resultBean = mockk<ResultBean>()
        every { resultBean.records } returns mData
        every { mobileDateUsageEntity.result } returns resultBean;

        val just: Observable<MobileDateUsageEntity> = Observable.just(mobileDateUsageEntity)

        every { model.getMobileDataUsage(HomePresenter.RESOURCE_ID, LIMIT) } returns just
        onGetMobileDataUsage = spyk()
        onGetDataError = spyk()
        showLoading = spyk()
        dismissLoading = spyk()
    }

    /**
     * 测试[HttpObserver] 配合[HomeView.onGetMobileDataUsage]
     * test [HttpObserver] with call [HomeView.onGetMobileDataUsage]
     */
    @Test
    fun testHomeViewCallSuccessWithHttpObservable() {
        log("┌─── testHomeViewCallSuccessWithHttpObservable ────────────────────────────────")

        val testObserver = TestObserver(httpObserver)
        model.getMobileDataUsage(HomePresenter.RESOURCE_ID, LIMIT)
                .subscribe(testObserver)
        testObserver.assertNoErrors()

        log("|\t▊ assert onGetMobileDataUsage called 1 , onGetDataError not call")
        verify(exactly = 1) { onGetMobileDataUsage.onGetMobileDataUsage() }
        verify { onGetDataError wasNot Called }
        log("└─── TEST PASS ──────────────────────────────────────────────────")
    }

    /**
     * 测试[HttpObserver] 配合[HomeView.onGetDataError]
     * test [HttpObserver] with call [HomeView.onGetDataError]
     */
    @Test
    fun testHomeViewCallFailureModelErrorWithHttpObservable() {
        log("┌─── testHomeViewCallFailureModelErrorWithHttpObservable ────────────────────────────────")

        val testObserver = TestObserver(httpObserver)
        model.getMobileDataUsage(HomePresenter.RESOURCE_ID, LIMIT)
                .map { t: MobileDateUsageEntity -> null }
                .subscribe(testObserver)
        testObserver.assertError(NullPointerException::class.java)

        log("|\t▊ assert onGetDataError called 1 , onGetMobileDataUsage not call")
        verify(exactly = 1) { onGetDataError.onGetDataError(any()) }
        verify { onGetMobileDataUsage wasNot Called }

        log("└─── TEST PASS ──────────────────────────────────────────────────")
    }

    /**
     * test 1:
     * 测试从 @see <a>https://data.gov.sg/api/action/datastore_search?resource_id=a807b7ab-6cad-4aa6-87d0-e283a7353a0f&limit=5</a> 获取数据并打印records
     * test 2:
     * 使用[com.witype.mvp.integration.UIObservableTransformer] 模拟界面的loading和dismiss
     *
     * test 1:
     * test get remote reource from
     * <a>https://data.gov.sg/api/action/datastore_search?resource_id=a807b7ab-6cad-4aa6-87d0-e283a7353a0f&limit=5</a> and print log
     * test 2:
     * use [com.witype.mvp.integration.UIObservableTransformer] mock UI loading and dismiss
     *
     */
    @Test
    fun testHttpObservableWithRemoteSuccess() {
        log("┌─── testHttpObservableWithRemoteSuccess ────────────────────────────────")
        val retrofit = mockRetrofit();

        val mobileDataUsage = retrofit.create(ApiModel::class.java)
                .getMobileDataUsage(HomePresenter.RESOURCE_ID, LIMIT)
        every { model.getMobileDataUsage(any(), any()) } returns mobileDataUsage

        val httpObserver = object : HttpObserver<RecordsBean>() {

            override fun onSuccess(t: RecordsBean) {
                //do something
            }

            override fun onError(e: Throwable) {
                log("|\t| onError => onGetDataError")
                onGetDataError.onGetDataError(e.message)
            }

            override fun onComplete() {
                log("|\t| onComplete => onGetMobileDataUsage")
                onGetMobileDataUsage.onGetMobileDataUsage()
            }
        }

        model.getMobileDataUsage(HomePresenter.RESOURCE_ID, LIMIT)
                .flatMap { t: MobileDateUsageEntity ->
                    log("|\t┌── response ────────────────────────────")
                    log("|\t| $t")
                    Observable.fromIterable(t.result?.records)
                }
                .doOnNext { t: RecordsBean? ->
                    log("|\t| $t")
                }
                .compose(bindUIEvent())
                .doOnComplete { log("|\t└─────────────────────────────────────────────────────") }
                .subscribe(TestObserver(httpObserver))
        log("|\t▊ assert onGetMobileDataUsage called 1 , onGetDataError not call")
        verify(exactly = 1) { onGetMobileDataUsage.onGetMobileDataUsage() }
        verify { onGetDataError wasNot Called }

        log("|\t▊ assert showLoading called 1 , dismissLoading call 1")
        verify(exactly = 1) { showLoading.showLoading() }
        verify(exactly = 1) { showLoading.dismissLoading() }

        log("|\t▊ assert call order showLoading => onGetMobileDataUsage => dismissLoading")
        verifySequence {
            showLoading.showLoading()
            showLoading.dismissLoading()
            onGetMobileDataUsage.onGetMobileDataUsage()
        }

        log("└─── TEST PASS ──────────────────────────────────────────────────")
    }

    /**
     * test 1:
     * 测试使用不完整的json数据获取[MobileDateUsageEntity]，并打印日志，模拟数据获取异常回调 [HomeView.onGetDataError]
     * test 2:
     * 使用[com.witype.mvp.integration.UIObservableTransformer] 模拟界面的loading和dismiss
     *
     * test 1:
     * test get [MobileDateUsageEntity] from snippet json and aspect call [HomeView.onGetDataError]
     * test 2:
     * use [com.witype.mvp.integration.UIObservableTransformer] mock UI loading and dismiss
     *
     */
    @Test
    fun testHttpObservableWithRemoteModelError() {
        log("┌─── testHttpObservableWithRemoteModelError ────────────────────────────────")

        //mock empty data
        val mobileDataUsage = Observable.just("{\"help\":\"https://data.gov.sg/api/3/action/help_show?name=datastore_search\",\"success\":true}")
                .flatMap { t ->
                    val data = Gson().fromJson(t, MobileDateUsageEntity::class.java)
                    Observable.just(data)
                }
        every { model.getMobileDataUsage(any(), any()) } returns mobileDataUsage

        val httpObserver = object : HttpObserver<RecordsBean>() {

            override fun onSuccess(t: RecordsBean) {
                //do something
            }

            override fun onError(e: Throwable) {
                super.onError(e)
                log("|\t| onError => onGetDataError")
                onGetDataError.onGetDataError(e.message)
            }

            override fun onComplete() {
                log("|\t| onComplete => onGetMobileDataUsage")
                onGetMobileDataUsage.onGetMobileDataUsage()
            }
        }
        val testObserver = TestObserver(httpObserver);
        model.getMobileDataUsage(HomePresenter.RESOURCE_ID, LIMIT)
                .flatMap { t: MobileDateUsageEntity ->
                    log("|\t┌── response size ${t.result?.records?.size ?: 0} ────────────────────────────")
                    log("|\t| $t")
                    Observable.fromIterable(t.result?.records)
                }
                .doOnNext { t: RecordsBean? ->
                    log("|\t| $t")
                }
                .compose(bindUIEvent())
                .doOnComplete { log("|\t└─────────────────────────────────────────────────────") }
                .subscribe(testObserver)
        testObserver.assertError(NullPointerException::class.java)

        log("|\t▊ assert onGetDataError called 1 , onGetMobileDataUsage not call")
        verify(exactly = 1) { onGetDataError.onGetDataError(any()) }
        verify { onGetMobileDataUsage wasNot Called }

        log("|\t▊ assert showLoading called 1 , dismissLoading call 1")
        verify(exactly = 1) { showLoading.showLoading() }
        verify(exactly = 1) { showLoading.dismissLoading() }

        log("|\t▊ assert call order showLoading => onGetDataError => dismissLoading")
        verifySequence {
            showLoading.showLoading()
            showLoading.dismissLoading()
            onGetDataError.onGetDataError(any())
        }
        log("└─── TEST PASS ──────────────────────────────────────────────────")
    }

    /**
     * 测试从 @see <a>https://data.gov.sg/api/action/datastore_search?resource_id=a807b7ab-6cad-4aa6-87d0-e283a7353a0f&limit=5</a> 获取数据
     * 然后通过[HomePresenter.DataRangeFilter] 、 [HomePresenter.DataGroupFun]、[HomePresenter.DataQuarterOffsetFun]过滤数据
     *
     * test get remote data from @see <a>https://data.gov.sg/api/action/datastore_search?resource_id=a807b7ab-6cad-4aa6-87d0-e283a7353a0f&limit=5</a>
     * then filter by [HomePresenter.DataRangeFilter] 、 [HomePresenter.DataGroupFun]、[HomePresenter.DataQuarterOffsetFun]
     *
     * ps full test in [com.witype.kotlindemo.DataFilterTest]
     */
    @Test
    fun testHttpObservableAndFunWithRemote() {
        log("┌─── testHttpObservableAndFunWithRemote ────────────────────────────────")

        val retrofit = mockRetrofit()
        val startAt = 2008;
        val endAt = 2018;

        val mobileDataUsage = retrofit.create(ApiModel::class.java)
                .getMobileDataUsage(HomePresenter.RESOURCE_ID, LIMIT)
        every { model.getMobileDataUsage(any(), any()) } returns mobileDataUsage

        val httpObserver = object : HttpObserver<List<RecordsBean>>() {

            override fun onSuccess(t: List<RecordsBean>) {
                //do something
            }

            override fun onError(e: Throwable) {
                super.onError(e)
                log("|\t| onError => onGetDataError")
                onGetDataError.onGetDataError(e.message)
            }

            override fun onComplete() {
                log("|\t| onComplete => onGetMobileDataUsage")
                onGetMobileDataUsage.onGetMobileDataUsage()
            }
        }

        val testObserver = TestObserver(httpObserver);
        model.getMobileDataUsage(HomePresenter.RESOURCE_ID, LIMIT)
                .map { t: MobileDateUsageEntity -> t.result?.records }
                .doOnNext { t: List<RecordsBean>? ->
                    log("|\t┌─── input ${t?.size} ────────────────────────────────")
                    t?.forEach { item: RecordsBean -> log("|\t| $item") }
                    log("|\t└─────────────────────────────────────────────")
                }
                .flatMap(HomePresenter.DataRangeFilter(startAt, endAt))
                .doOnNext { t: List<RecordsBean>? ->
                    log("|\t┌─── after filter ${t?.size} ────────────────────────────────")
                    t?.forEach { item: RecordsBean ->
                        log("|\t| $item")
                    }
                    log("|\t└─────────────────────────────────────────────")
                }
                .flatMap(HomePresenter.DataGroupFun(null))
                .doOnNext { t: List<RecordsBean>? ->
                    log("|\t┌─── group and calc item ${t?.size} ────────────────────────────────")
                    log("|\t| ▊ group data ${t?.size} ")
                    t?.forEach { item: RecordsBean -> log("|\t| $item") }
                }
                .flatMap(HomePresenter.DataQuarterOffsetFun())
                .doOnNext { t: List<RecordsBean>? ->
                    log("|\t| ▊ after calc ${t?.size} ")
                    log("|\t| quarter \t\t\t| usage \t\t\t| offset \t\t\t| total \t\t\t")
                    t?.forEach { item: RecordsBean ->
                        log("|\t| ${item.quarter} \t\t\t| ${item.volume_of_mobile_data} \t\t\t| ${item.volume_offset.toFloat().toString()} \t\t\t| ${item.total_of_year} \t\t\t")
                    }
                    log("|\t└─────────────────────────────────────────────")
                }
                .compose(bindUIEvent())
                .subscribe(testObserver)
        testObserver.assertNoErrors()

        log("|\t▊ assert onGetMobileDataUsage called 1 , onGetDataError not call")
        verify(exactly = 1) { onGetMobileDataUsage.onGetMobileDataUsage() }
        verify { onGetDataError wasNot Called }

        log("|\t▊ assert showLoading called 1 , dismissLoading call 1")
        verify(exactly = 1) { showLoading.showLoading() }
        verify(exactly = 1) { showLoading.dismissLoading() }

        log("|\t▊ assert call order showLoading => onGetMobileDataUsage => dismissLoading")
        verifySequence {
            showLoading.showLoading()
            showLoading.dismissLoading()
            onGetMobileDataUsage.onGetMobileDataUsage()
        }

        log("|\t*** DataRangeFilter 、DataGroupFun、DataQuarterOffsetFun testcase see DataFilterTest ***")

        log("└─── TEST PASS ──────────────────────────────────────────────────")
    }

    private fun mockRetrofit(): Retrofit {
        return Retrofit.Builder().baseUrl("https://data.gov.sg")
                .client(OkHttpClient.Builder().build())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .addConverterFactory(GsonConverterFactory.create())
                .build()
    }

    val httpObserver = object : HttpObserver<MobileDateUsageEntity>() {

        override fun onSuccess(t: MobileDateUsageEntity) {
            log("|\t| onSuccess => onGetMobileDataUsage")
            onGetMobileDataUsage.onGetMobileDataUsage()
        }

        override fun onError(e: Throwable) {
            log("|\t| onError => onGetDataError")
            onGetDataError.onGetDataError(e.message)
        }
    }

    fun <T> bindUIEvent(): UIObservableTransformer<T> {
        return object : UIObservableTransformer<T>() {
            override fun hasSubscribe() {
                log("|\t| hasSubscribe => showLoading")
                showLoading.showLoading()
            }

            override fun willComplete() {
                log("|\t| willComplete => dismissLoading")
                showLoading.dismissLoading()
            }
        }
    }

    private fun log(message: String) {
        println(message)
    }

    @After
    fun done() {
        RxAndroidPlugins.reset()
        RxJavaPlugins.reset();
    }
}